### PR TITLE
Allow providing an OAuth token for authenticating against k8s

### LINKED ISSUE #179: Support using an OAuth token for authenticating against the API server

### PR DESCRIPTION
**Pull request by mccheah**
Closes #179

### CODE REVIEW

**Comment by mccheah** [comment]
@foxish  @robert3005 for review.

I don't quite know how to test this. Does Minikube support authentication with OAuth tokens instead of certificate and key files? I need this for an internal application but it seems like a good idea to have this option in general.

The other question is whether or not the token contents should be provided directly in the Spark configuration or if the token configurations should be passed through a file. I chose the latter because it seems more secure. The `SparkConf` is presented on the Spark UI page and can be logged if the user sets `spark.logConf` to `true`.

**Comment by robert3005** [comment]
Hmm, setting this on spark conf sounds like asking for trouble to me. I am not certain if there's a better place though. There's some sanitisation available in current release but the users need to enable it.

**Comment by mccheah** [comment]
Well in this change we're not passing the token itself, but only the location on the submitter's disk. That way the token itself isn't ever logged anywhere.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala [hunk 1]
```diff
@@ -83,6 +83,18 @@ package object config {
...
+    ConfigBuilder("spark.kubernetes.submit.oauthTokenFile")
+      .doc("""
+             | File containing an OAuth token for authenticating
+             | against the Kubernetes API server. This file
+             | should be located on the submitting machine's
+             | disk.
+           """.stripMargin)
+      .stringConf
+      .createOptional
+
```
Comment (robert3005) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: remove whitespace

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 2]
```diff
@@ -131,6 +132,13 @@ private[spark] class Client(
     sparkConf.get(KUBERNETES_CLIENT_CERT_FILE).foreach {
       f => k8ConfBuilder = k8ConfBuilder.withClientCertFile(f)
     }
+    sparkConf.get(KUBERNETES_OAUTH_TOKEN_FILE).foreach { f =>
+      val oauthTokenFile = new File(f)
+      require(oauthTokenFile.isFile, s"OAuth token file provided at $f does not exist or is" +
```
Comment (robert3005) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: Can we move the whole message to the new line and not break it?

**Review by robert3005 — COMMENTED** [review]
Can we maybe send the file across to the pods maybe instead of sparkconf?

**Comment by robert3005** [comment]
Right, sorry I misread some other comment. Looks good now

**Comment by mccheah** [comment]
The more I think about it the more I think it makes sense to be allowed to pass the token value via Spark configuration instead of through a file. The reason is that it allows users of the `SparkLauncher` API to avoid having to push the token data down to disk.

We can erase the token value from the Spark configuration after we use it.

**Comment by mccheah** [comment]
@foxish is this reasonable?

**Comment by foxish** [comment]
@mccheah, the fabric8 kubernetes-client has support for using OAuth tokens instead of client-certs which is what we use against GKE, if the .kube/config file is setup appropriately. Do we need this explicitly as an option to Spark submit if fabric8 handles it for us?

**Comment by mccheah** [comment]
It is useful to be able to provide the token dynamically instead of relying on a static configuration file. Multiple applications being submitted from the same host but from different users is an example.

**Comment by foxish** [comment]
My .kube/config contains many clusters and a mechanism by which I am authorized to talk to them. The way I switch which cluster to talk to is using https://kubernetes.io/docs/user-guide/kubectl/kubectl_config_use-context/. If we expect a user to get stdout/stderr using `kubectl get logs` from the driver pod, they do need to switch to the context in which that SparkJob is running. In that case, do we gain a lot from this? It's a minor change however, so, if you have a good reason (like automation around job submission to different clusters), I'm okay with this change.

**Comment by mccheah** [comment]
One example use case is that the application submitter does not itself have the credentials, but rather is provided the credentials by some other party and the submitter wishes to forward those credentials through to spark-submit.

**Comment by foxish** [comment]
That sounds reasonable. LGTM.

**Comment by mccheah** [comment]
@robert3005 good to merge?

**Comment by robert3005** [comment]
Yup.

**Pull request merged**